### ISSUE TITLE
How to use liquid names with Spaces?

### ISSUE THREAD

**Issue by Voidi** [stanhebben/MineTweaker3#150]
Ethanol from Rotarycraft has ``<liquid:rc ethanol>`` as internal name but used in a script i get:
``Could not resolve <liquid : rc ethanol>``

**Comment by UndeadZeratul** [comment]
I've had the same issue with Project Red and their fluid names, as shown [here](https://github.com/MrTJP/ProjectRed/issues/705).  If MineTweaker can add support for spaces, that'd work great, but I also wonder if the mods that have spaces in their names could just remove them...

Either way, it's a problem for sure.

**Comment by Belgabor** [comment]
Same issue with Blood Magic Life Essence (`<liquid:life essence>`).

**Comment by Mrbysco** [comment]
I just asked in the bloodmagic channel if he could change life essence to something like life_essence or life.essence.. Ill see what happens..

**Comment by jaredlll08** [comment]
Not our fault, nothing we can do, you should NEVER use a space in an unlocalized name.

**Comment by Voidi** [comment]
@jaredlll08 i know using space is bad practice, but a lot of mods uses it. Is impossible to implement?

**Comment by stanhebben** [comment]
It's possible to implement (I did it for items), just a bit of work to do so.

**Comment by jaredlll08** [comment]
really? where can I find that? I thought it was impossible because you can't do `/give @a testmod:saw red`

**Comment by stanhebben** [comment]
It's not possible in command line but it is possible in scripts. It essentially involves stripping spaces from all registered items and building an index with that. See https://github.com/stanhebben/MineTweaker3/blob/master/MineTweaker3-MC1710-Main/src/main/java/minetweaker/mc1710/brackets/ItemBracketHandler.java#L44

**Comment by jaredlll08** [comment]
ah

**Comment by stanhebben** [comment]
Essentially the bracket handler (for which there is a version for items, liquids and ore dictionary entries) receives a list of tokens. Concatenating the tokens gives the item names.

However, the whitespace between tokens is not recorded in the tokener, so we have to work around that by building an index without whitespace.

**Comment by Yulife** [comment]
@jaredlll08 don't close issues so hastily! lol Often *your* impossible is possible ;)

**Comment by Pablode** [comment]
> It's possible to implement (I did it for items), just a bit of work to do so.

Would be awesome to see this fixed. :)

**Comment by Pablode** [comment]
As I understand it, you simply have to implement an internal map of fluid names as done in [ItemBracketHandler](https://github.com/stanhebben/MineTweaker3/blob/3e33904eddc796d183047f28d8be815194e634ad/MineTweaker3-MC1710-Main/src/main/java/minetweaker/mc1710/brackets/ItemBracketHandler.java) into the [FluidBracketHandler](https://github.com/stanhebben/MineTweaker3/blob/8d7620c92e250326b641f25bb709be18c18c3515/MineTweaker3-MC1710-Main/src/main/java/minetweaker/mc1710/brackets/LiquidBracketHandler.java). This does seem like a relatively easy task.

**Comment by Voidi** [comment]
Wasn't this already solved?

**Comment by Blue64** [comment]
Yes, but they want an easier way beyond mapping out our entire fluid registry, striping it of white space in an internal index and implementing it that way. I'm not sure I'm looking forward to this task for adding the Liquid Dictionary of Liquifacted Coal (Thermal Foundation) to my furnace burn time (16k, standard Coal Block burn)

**Comment by Blue64** [comment]
Also, is there a way to write stuff to a text file? it'd make this task much easier for sure. If I can see these fluid registries, I can work with em.

**Comment by Blue64** [comment]
@Mrbysco it would appear they still haven't fixed Life Essence, did they ever mention anything after you mentioned it to them?

@stanhebben @Pablod how exactly would I accomplish this? I'm having a difficult enough time understanding the tutorial (which I might add doesn't explain very much beyond the basics as far as I can tell), this seems a few steps beyond what information I've found, and I've got zero clue how to implement it on my end (not for a lack of trying though). The links help, but I've still got no clue how to utilize this information, as when I start speaking Java in a ZenScript file, it throws error after error at me, and sometimes even speaks in Assembly (thank you Cheat Engine for that comprehension), which I understand as just strings of base 256 code, I have no clue how to actually figure out what it's trying to tell me.
